test/emulator_backend.go

```diff
@@ -493,3 +493,31 @@ func (e *EmulatorBackend) replaceImports(code string) string {
 func (e *EmulatorBackend) StandardLibraryHandler() stdlib.StandardLibraryHandler {
 	return e.stdlibHandler
 }
+
+// GetEvents returns all the emitted events up until the latest block,
+// optionally filtered by event type.
+func (e *EmulatorBackend) GetEvents(eventType string) []cadence.Event {
+	latestBlock, err := e.blockchain.GetLatestBlock()
+	if err != nil {
+		panic(err)
+	}
+
+	latestBlockHeight := latestBlock.Header.Height
+	height := uint64(0)
+	allEvents := make([]cadence.Event, 0)
+	for height <= latestBlockHeight {
+		events, err := e.blockchain.GetEventsByHeight(
+			height,
+			eventType,
+		)
+		if err != nil {
+			panic(err)
+		}
+		for _, event := range events {
+			allEvents = append(allEvents, event.Value)
+		}
+		height += 1
+	}
+
+	return allEvents
```

Comment:
> Like I suggested in the other PR https://github.com/onflow/cadence/pull/2505#discussion_r1205900239, we can convert the `[]cadence.Event` to `[]interpreter.Value` here, and return that.

Thanks for the valuable feedback :pray: It was really helpful to unblock me :bow: I have added a first version here: https://github.com/onflow/cadence-tools/pull/113/commits/72eee50a3699d94abcc474141d6eec48668dfbcc

> btw, despite where the conversion happens, I think we would run into the problem we discussed earlier, where the Event types defined in the blockchain are not available at this point. Or did you manage to workaround that?

You are right, we still need to tackle that issue. However, this is something that we'll have to solve anyway. That's why I am adding this functionality a priori, even though it might not be functional for the time being.

Reply:
Bastian had mentioned:

> I think it involves hooking up the checker's/interpeter's import location handler

But I haven't taken any steps yet on trying this lead.

Reply:
Would it be possible to somehow do `runtime.ExportValue` for run-time types that live on the `EmulatorBackend`'s `Blockchain` ? All these types must live on `common.AddressLocation`, such as `"A.01cf0e2f2f715450.FooContract.NumberAdded"`

Reply:
> That's why I am adding this functionality a priori, even though it might not be functional for the time being.

It would be a bit bad for UX, to provide an API that is broken by default. e.g.: Users will see there's an API to get events, but when they use it they'll start complaining about the API being broken, etc.

It might be better to hold off this API until we find a solution for types.

Reply:
I'll also have a look today/early next week, on this a bit. Maybe we can also have a quick brainstorming session if needed.

Reply:
That would be great :pray: